preheaders/checkout.php

```diff
@@ -562,242 +567,40 @@
-					$blastname,
-				) );
-			}
-
-			// Check if billing details are available, if not adjust the arrays.
-			if ( ! empty( $baddress1 ) ) {
-				$meta_keys = array_merge( $meta_keys, array(
-					"pmpro_baddress1",
-					"pmpro_baddress2",
-					"pmpro_bcity",
-					"pmpro_bstate",
-					"pmpro_bzipcode",
-					"pmpro_bcountry",
-					"pmpro_bphone",
-					"pmpro_bemail",
-				) );
-
-				$meta_values = array_merge( $meta_values, array(
-					$baddress1,
-					$baddress2,
-					$bcity,
-					$bstate,
-					$bzipcode,
-					$bcountry,
-					$bphone,
-					$bemail,
-				) );
-			}
-
-			pmpro_replaceUserMeta( $user_id, $meta_keys, $meta_values );
-
-			//save first and last name fields
-			if ( ! empty( $bfirstname ) ) {
-				$old_firstname = get_user_meta( $user_id, "first_name", true );
-				if ( empty( $old_firstname ) ) {
-					update_user_meta( $user_id, "first_name", $bfirstname );
-				}
-			}
-			if ( ! empty( $blastname ) ) {
-				$old_lastname = get_user_meta( $user_id, "last_name", true );
-				if ( empty( $old_lastname ) ) {
-					update_user_meta( $user_id, "last_name", $blastname );
-				}
-			}
-
-			if( $pmpro_level->expiration_period == 'Hour' ){
-				update_user_meta( $user_id, 'pmpro_disable_notifications', true );
-			}
-
-			//show the confirmation
-			$ordersaved = true;
-
-			//hook
-			do_action( "pmpro_after_checkout", $user_id, $morder );    //added $morder param in v2.0
-
-			$sendemails = apply_filters( "pmpro_send_checkout_emails", true);
-
-			if($sendemails) { // Send the emails only if the flag is set to true
+	// Make sure we have an order object.
+	if ( empty( $morder ) ) {
+		$morder                 = new MemberOrder();
+		$morder->InitialPayment = 0;
+		$morder->Email          = $bemail;
+		$morder->gateway        = 'free';
+		$morder->status			= 'success';
+		$morder = apply_filters( "pmpro_checkout_order_free", $morder );
+	}
 
-				//setup some values for the emails
-				if ( ! empty( $morder ) ) {
-					$order = new MemberOrder( $morder->id );
-				} else {
-					$order = null;
-				}
-				$current_user->membership_level = $pmpro_level; //make sure they have the right level info
+	//add an item to the history table, cancel old subscriptions
+	if ( ! empty( $morder ) ) {
+		$morder->user_id       = $user_id;
+		$morder->membership_id = $pmpro_level->id;
+		$morder->saveOrder();
+	}
 
-				//send email to member
-				$pmproemail = new PMProEmail();
-				$pmproemail->sendCheckoutEmail( $current_user, $order );
+	if ( pmpro_complete_async_checkout( $morder ) ) {
+		//redirect to confirmation
+		$rurl = pmpro_url( "confirmation", "?pmpro_level=" . $pmpro_level->id );
+		$rurl = apply_filters( "pmpro_confirmation_url", $rurl, $user_id, $pmpro_level );
+		wp_redirect( $rurl );
+		exit;
+	} else {
 
-				//send email to admin
-				$pmproemail = new PMProEmail();
-				$pmproemail->sendCheckoutAdminEmail( $current_user, $order );
-			}
+		//uh oh. we charged them then the membership creation failed

```

Comment:
It's not clear to me here without looking inside pmpro_complete_checkout() that an error during "membership creation" is the only failure state.

This is all meant as an in progress refactor, so I don't think we need to change any of the code, but perhaps update the comment here to be more clear what pmpro_complete_checkout() is doing and why we are doing the tests we are below. i.e. if it returned false, we know user creation failed.

Reply:
Users are always created before pmpro_complete_checkout() is called. pmpro_complete_chekcout() returns true if `pmpro_changeMembershipLevel( $custom_level, $order->user_id, 'changed' )` completes successfully, false otherwise. So in the pmpro_msg, "membership creation" is assigning a membership level. Maybe we need clearer wording in the error message

Reply:
Maybe it would be better to say "Something went wrong while processing your checkout."

Reply:
Improved the comment and error message